lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
 };
 } // end anonymous namespace
 
+namespace {
+class JVPEmitter final
+    : public TypeSubstCloner<JVPEmitter, SILOptFunctionBuilder> {
+private:
+  /// The global context.
+  ADContext &context;
+
+  /// The original function.
+  SILFunction *const original;
+
+  /// The `[differentiable]` attribute.
+  SILDifferentiableAttr *const attr;
+
+  /// The JVP function.
+  SILFunction *const jvp;
+
+  /// The differential function.
+  SILFunction *differential;
+
+  /// The differentiation invoker.
+  DifferentiationInvoker invoker;
+
+  /// Caches basic blocks whose phi arguments have been remapped (adding a
+  /// predecessor enum argument).
+  SmallPtrSet<SILBasicBlock *, 4> remappedBasicBlocks;
+
+  /// Temporary boolean for determine whether we should be generating the body of the JVP.
+  /// If true, then skip the JVP generation process since we cannot generate JVPs for the
+  /// primitive VJPs defined on addition for Float for example.
+  bool vjpGenerated;
```

Comment:
It would be nicer if this is checked by the caller of `JVPEmitter` instead of being leaked into `JVPEmitter`

Reply:
I saw `bool JVPEmitter::run()` as the main function that creates a function, which in this case is called from `processDifferentiableAttribute`. Are you suggesting to add two different `run()` functions, one for if the VJP was generated and another if it wasn't? So I can then call either one of them in `processDifferentiableAttribute`?

Reply:
Sorry, I meant to suggest not creating a `JVPEmitter` at all for JVPs that are supposed to return `undef`.

Reply:
Ahh I see sounds good!